helix-view/src/modeline.rs

```diff
@@ -0,0 +1,258 @@
+use std::borrow::Cow;
+
+use helix_core::indent::IndentStyle;
+use helix_core::regex::Regex;
+use helix_core::{LineEnding, Rope};
+use once_cell::sync::Lazy;
+
+// 5 is the vim default
+const LINES_TO_CHECK: usize = 5;
+const LENGTH_TO_CHECK: usize = 256;
+
+static MODELINE_REGEX: Lazy<Regex> =
+    Lazy::new(|| Regex::new(r"(?:\S+\s+)?(?:vi|vim|Vim|ex):\s*(?:set? )?(.*)(?::[^:]*)?").unwrap());
+
+#[derive(Default, Debug, Eq, PartialEq)]
+pub struct Modeline {
+    language: Option<String>,
+    indent_style: Option<IndentStyle>,
+    line_ending: Option<LineEnding>,
+}
+
+impl Modeline {
+    pub fn parse(text: &Rope) -> Self {
+        let mut modeline = Self::default();
+
+        for line in text.lines().take(LINES_TO_CHECK).chain(
+            text.lines_at(text.len_lines())
+                .reversed()
+                .take(LINES_TO_CHECK),
+        ) {
+            // can't guarantee no extra copies, since we need to regex and
+            // regexes can't operate over chunks yet, but we can at least
+            // limit how much we potentially need to copy because modelines
+            // are typically quite short.
+            let line = Cow::<str>::from(line.slice(..(LENGTH_TO_CHECK.min(line.len_chars()))));
+            if let Some(line_modeline) = Self::parse_line(&line) {
+                modeline.merge(line_modeline);
+            }
+        }
+
+        modeline
+    }
+
+    pub fn language(&self) -> Option<&str> {
+        self.language.as_deref()
+    }
+
+    pub fn indent_style(&self) -> Option<IndentStyle> {
+        self.indent_style
+    }
+
+    pub fn line_ending(&self) -> Option<LineEnding> {
+        self.line_ending
+    }
+
+    fn parse_line(line: &str) -> Option<Self> {
+        if let Some(captures) = MODELINE_REGEX.captures(line) {
+            let mut language = None;
+            let mut indent_style = None;
+            let mut line_ending = None;
+
+            let options = &captures[1];
+            for option in options.split([' ', '\t', ':']) {
```

Comment:
Using a simple split here doesn't work since mode lines allow escaping `with \:` you need to handle that here. You can take a look at the `query_atom` function to see how you can use split to split on a char that can be escaped (in that case it was spaces)